File: test/kms-stub.mjs
/*
 * A local stand-in for the KMS API, spoken over the real wire protocol so the real
 * AWS SDK talks to it unmodified. That covers the actual client, signing, endpoint
 * resolution and error unmarshalling rather than a shortcut around them.
 *
 * The contract below was verified against a real Aws::KMS::KMSClient:
 *   POST /  Content-Type: application/x-amz-json-1.1
 *   X-Amz-Target: TrentService.GetPublicKey | TrentService.Sign
 *   blobs base64, errors as {"__type": "...", "message": "..."}
 * Plain http:// is fine -- no TLS anywhere -- and the SDK never verifies a
 * response signature. It does refuse to send with no credentials at all, so the
 * caller must set dummy ones.
 *
 * Signing goes through the `openssl` CLI rather than node:crypto because KMS's
 * MessageType semantics map exactly onto pkeyutl and not onto anything node
 * exposes: there is no node API for "sign this already-computed digest", and none
 * at all for ML-DSA external-mu.
 */
import { createServer } from 'node:http';
import { generateKeyPairSync } from 'node:crypto';
import { spawnSync } from 'node:child_process';
import { mkdtempSync, writeFileSync, readFileSync, rmSync } from 'node:fs';
import { tmpdir } from 'node:os';
import { join } from 'node:path';

const OPENSSL = process.env.AWSKMS_OPENSSL ?? 'openssl';

/* Which KeySpecs this stub can serve depends on the openssl CLI it has. */
function opensslSupports(algorithm) {
  const r = spawnSync(OPENSSL, ['genpkey', '-algorithm', algorithm, '-out', '/dev/null'], {
    stdio: 'ignore',
  });
  return r.status === 0;
}

const SPECS = {
  RSA_2048: { algorithm: 'rsa', options: { modulusLength: 2048 } },
  RSA_3072: { algorithm: 'rsa', options: { modulusLength: 3072 } },
  RSA_4096: { algorithm: 'rsa', options: { modulusLength: 4096 } },
  ECC_NIST_P256: { algorithm: 'ec', options: { namedCurve: 'prime256v1' } },
  ECC_NIST_P384: { algorithm: 'ec', options: { namedCurve: 'secp384r1' } },
  ECC_NIST_P521: { algorithm: 'ec', options: { namedCurve: 'secp521r1' } },
  ECC_SECG_P256K1: { algorithm: 'ec', options: { namedCurve: 'secp256k1' } },
  ECC_NIST_EDWARDS25519: { algorithm: 'ed25519', options: {} },
  ML_DSA_44: { algorithm: 'ml-dsa-44', options: {}, cliName: 'ML-DSA-44' },
  ML_DSA_65: { algorithm: 'ml-dsa-65', options: {}, cliName: 'ML-DSA-65' },
  ML_DSA_87: { algorithm: 'ml-dsa-87', options: {}, cliName: 'ML-DSA-87' },
};

const SIGNING_ALGORITHMS = {
  RSA: [
    'RSASSA_PSS_SHA_256',
    'RSASSA_PSS_SHA_384',
    'RSASSA_PSS_SHA_512',
    'RSASSA_PKCS1_V1_5_SHA_256',
    'RSASSA_PKCS1_V1_5_SHA_384',
    'RSASSA_PKCS1_V1_5_SHA_512',
  ],
  ECC_NIST_P256: ['ECDSA_SHA_256'],
  ECC_NIST_P384: ['ECDSA_SHA_384'],
  ECC_NIST_P521: ['ECDSA_SHA_512'],
  ECC_SECG_P256K1: ['ECDSA_SHA_256'],
  ECC_NIST_EDWARDS25519: ['ED25519_SHA_512', 'ED25519_PH_SHA_512'],
  ML_DSA: ['ML_DSA_SHAKE_256'],
};

const signingAlgorithmsFor = (spec) =>
  SIGNING_ALGORITHMS[spec] ??
  (spec.startsWith('RSA_') ? SIGNING_ALGORITHMS.RSA : SIGNING_ALGORITHMS.ML_DSA);

const DIGEST_OF = { SHA_256: 'sha256', SHA_384: 'sha384', SHA_512: 'sha512' };

/* HTTP status only affects the SDK's retry classification; the exception name is
 * what it maps to an error type. Values from the KMS API reference. */
const STATUS = {
  NotFoundException: 400,
  DisabledException: 400,
  InvalidKeyUsageException: 400,
  KMSInvalidStateException: 400,
  UnsupportedOperationException: 400,
  ValidationException: 400,
  SerializationException: 400,
  AccessDeniedException: 400,
  KMSInternalException: 500,
  KeyUnavailableException: 500,
  DependencyTimeoutException: 500,
  ThrottlingException: 400,
};

export function createKmsStub({ verbose = !!process.env.AWSKMS_STUB_VERBOSE } = {}) {
  const dir = mkdtempSync(join(tmpdir(), 'awskms-stub-'));
  const keys = new Map(); // key id -> { spec, pemPath, spkiDer }
  let requests = [];

  /* Any KeySpec name appearing in the key id selects that spec, so tests can use
   * realistic aliases and ARNs. Generated on first use and kept for the life of
   * the stub, so load/sign/verify all agree. */
  function resolve(keyId) {
    if (keys.has(keyId)) return keys.get(keyId);
    const spec = Object.keys(SPECS).find((s) =>
      keyId.toLowerCase().includes(s.toLowerCase()),
    );
    if (!spec) return null;
    const { algorithm, options } = SPECS[spec];
    let pair;
    try {
      pair = generateKeyPairSync(algorithm, options);
    } catch {
      return null; // host openssl too old for this spec
    }
    const pemPath = join(dir, `${spec}-${keys.size}.pem`);
    writeFileSync(pemPath, pair.privateKey.export({ type: 'pkcs8', format: 'pem' }));
    const record = {
      spec,
      pemPath,
      spkiDer: pair.publicKey.export({ type: 'spki', format: 'der' }),
    };
    keys.set(keyId, record);
    return record;
  }

  const send = (res, status, body) => {
    const buf = Buffer.from(JSON.stringify(body));
    res.writeHead(status, {
      'content-type': 'application/x-amz-json-1.1',
      'content-length': buf.length,
      'x-amzn-RequestId': 'stub-00000000-0000-0000-0000-000000000000',
    });
    res.end(buf);
  };
  const fail = (res, type, message) =>
    send(res, STATUS[type] ?? 400, { __type: type, message });

  /*
   * A key id containing one of these markers makes GetPublicKey return something
   * malformed, which is what exercises the provider's own validation. Those paths
   * are otherwise unreachable offline: they guard against a KeySpec this provider
   * does not implement, and against a SubjectPublicKeyInfo disagreeing with the
   * KeySpec reported beside it, both producible only by the real service.
   *
   * The markers match those in src/kms_stub.c, so the same tests run against both
   * backends.
   */
  function injectFault(keyId, rec) {
    let { spec, spkiDer } = rec;
    /* A key that exists and parses fine but is not for signing. Only the real
     * backend checks this -- the offline one has no KeyUsage concept -- so it is
     * the HTTP stub that can reach kms_aws.cc's guard, through the real SDK. */
    let keyUsage = keyId.includes('fault-keyusage') ? 'ENCRYPT_DECRYPT' : 'SIGN_VERIFY';

    if (keyId.includes('fault-badspec')) spec = 'SM2';
    else if (keyId.includes('fault-wrongtype')) spec = 'RSA_2048';
    else if (keyId.includes('fault-wronggroup')) {
      spec = spec === 'ECC_NIST_P256' ? 'ECC_NIST_P384' : 'ECC_NIST_P256';
    }

    if (keyId.includes('fault-emptyspki')) spkiDer = Buffer.alloc(0);
    else if (keyId.includes('fault-truncspki')) {
      spkiDer = spkiDer.subarray(0, Math.floor(spkiDer.length / 2));
    } else if (keyId.includes('fault-badspki')) {
      // Corrupt the body but keep the leading SEQUENCE tag, so this fails in the
      // parser rather than being rejected as obviously-not-DER.
      spkiDer = Buffer.from(spkiDer);
      for (let i = spkiDer.length >> 1; i < spkiDer.length; i++) spkiDer[i] ^= 0xff;
    }

    return { spec, spkiDer, keyUsage };
  }

  /*
   * `fault-err-<ExceptionName>` in a key id makes GetPublicKey return that KMS
   * exception; `fault-signerr-<ExceptionName>` lets the key load and fails the
   * Sign instead.
   *
   * These exercise reason_for() in src/kms_aws.cc, the table turning a KMS
   * exception into the err.code a caller sees. The request goes through the real
   * AWS SDK, so they also pin down the SDK's own classification: whether
   * aws-sdk-cpp maps the wire string "DisabledException" onto KMSErrors::DISABLED
   * is a fact about the SDK, and this is the only place it is checked.
   */
  function requestedError(keyId, prefix) {
    const m = new RegExp(`${prefix}-([A-Za-z]+Exception)`).exec(keyId);
    return m?.[1] ?? null;
  }

  function getPublicKey(res, body) {
    const keyId = body.KeyId ?? '';
    const requested = requestedError(keyId, 'fault-err');
    if (requested) return fail(res, requested, `stub: injected ${requested}`);
    const rec = resolve(keyId);
    if (!rec) return fail(res, 'NotFoundException', `Key '${keyId}' does not exist`);
    const { spec, spkiDer, keyUsage } = injectFault(keyId, rec);
    return send(res, 200, {
      KeyId: `arn:aws:kms:us-east-1:000000000000:key/${keyId}`,
      KeySpec: spec,
      KeyUsage: keyUsage,
      SigningAlgorithms: signingAlgorithmsFor(rec.spec),
      PublicKey: spkiDer.toString('base64'),
    });
  }

  /* Mirrors KMS's MessageType semantics exactly, via pkeyutl. */
  function pkeyutlArgs(rec, algorithm, messageType) {
    const args = ['pkeyutl', '-sign', '-inkey', rec.pemPath];

    if (algorithm === 'ML_DSA_SHAKE_256') {
      if (messageType !== 'EXTERNAL_MU') return { error: 'only EXTERNAL_MU is implemented for ML-DSA' };
      // The input already IS mu, which is what mu:1 means.
      return { args: [...args, '-rawin', '-pkeyopt', 'mu:1'] };
    }
    if (algorithm.startsWith('ED25519')) {
      if (messageType !== 'RAW') return { error: 'ED25519_SHA_512 requires MessageType RAW' };
      return { args: [...args, '-rawin'] };
    }
    if (messageType !== 'DIGEST') {
      return { error: `the stub only implements MessageType DIGEST for ${algorithm}` };
    }
    if (algorithm.startsWith('ECDSA_')) {
      const md = DIGEST_OF[algorithm.slice('ECDSA_'.length)];
      return { args: [...args, '-pkeyopt', `digest:${md}`] };
    }
    if (algorithm.startsWith('RSASSA_PSS_')) {
      const md = DIGEST_OF[algorithm.slice('RSASSA_PSS_'.length)];
      // KMS always uses salt length == digest length, and MGF1 == the digest.
      return {
        args: [
          ...args,
          '-pkeyopt', 'rsa_padding_mode:pss',
          '-pkeyopt', 'rsa_pss_saltlen:digest',
          '-pkeyopt', `digest:${md}`,
        ],
      };
    }
    if (algorithm.startsWith('RSASSA_PKCS1_V1_5_')) {
      const md = DIGEST_OF[algorithm.slice('RSASSA_PKCS1_V1_5_'.length)];
      return {
        args: [...args, '-pkeyopt', 'rsa_padding_mode:pkcs1', '-pkeyopt', `digest:${md}`],
      };
    }
    return { error: `unknown SigningAlgorithm ${algorithm}` };
  }

  function sign(res, body) {
    const keyId = body.KeyId ?? '';
    /* The key loaded fine; the Sign is what fails. This is the path that matters
     * most, because it is the one a long-running process hits after a key that
     * worked at startup is disabled or scheduled for deletion underneath it. */
    const requested = requestedError(keyId, 'fault-signerr');
    if (requested) return fail(res, requested, `stub: injected ${requested}`);

    const rec = resolve(keyId);
    if (!rec) return fail(res, 'NotFoundException', `Key '${body.KeyId}' does not exist`);

    const algorithm = body.SigningAlgorithm;
    const messageType = body.MessageType ?? 'RAW';
    if (!signingAlgorithmsFor(rec.spec).includes(algorithm)) {
      return fail(
        res,
        'InvalidKeyUsageException',
        `Algorithm ${algorithm} is incompatible with key spec ${rec.spec}`,
      );
    }

    const message = Buffer.from(body.Message ?? '', 'base64');
    // KMS's documented Message constraint. Ed25519 is the only spec that feels
    // it, and the provider is expected to reject oversize input before it gets
    // here -- so if this fires from a test, the provider has a bug.
    if (message.length < 1 || message.length > 4096) {
      return fail(res, 'ValidationException', 'Message must be 1-4096 bytes');
    }
    if (messageType === 'EXTERNAL_MU' && message.length !== 64) {
      return fail(res, 'ValidationException', 'EXTERNAL_MU Message must be 64 bytes');
    }

    const plan = pkeyutlArgs(rec, algorithm, messageType);
    if (plan.error) return fail(res, 'UnsupportedOperationException', plan.error);

    const inPath = join(dir, 'in.bin');
    const outPath = join(dir, 'out.bin');
    writeFileSync(inPath, message);
    const r = spawnSync(OPENSSL, [...plan.args, '-in', inPath, '-out', outPath]);
    if (r.status !== 0) {
      return fail(
        res,
        'KMSInternalException',
        `stub signing failed: ${(r.stderr ?? '').toString().split('\n')[0]}`,
      );
    }

    return send(res, 200, {
      KeyId: `arn:aws:kms:us-east-1:000000000000:key/${body.KeyId}`,
      Signature: readFileSync(outPath).toString('base64'),
      SigningAlgorithm: algorithm,
    });
  }

  const server = createServer((req, res) => {
    const chunks = [];
    req.on('data', (c) => chunks.push(c));
    req.on('end', () => {
      const raw = Buffer.concat(chunks).toString('utf8');
      const target = req.headers['x-amz-target'] ?? '';
      let body;
      try {
        body = raw ? JSON.parse(raw) : {};
      } catch {
        return fail(res, 'SerializationException', 'malformed JSON body');
      }
      // The introspection endpoints are not part of the KMS API and must not be
      // recorded, or reading the log would change it. They let a test running in
      // a child process see what the provider actually asked for, which is how
      // "a size probe costs nothing" is asserted.
      if (req.url === '/__requests') {
        if (req.method === 'GET') return send(res, 200, { requests });
        if (req.method === 'DELETE') {
          requests = [];
          return send(res, 200, { ok: true });
        }
      }

      requests.push({ target, keyId: body.KeyId, messageType: body.MessageType,
                      signingAlgorithm: body.SigningAlgorithm,
                      messageLength: body.Message ? Buffer.from(body.Message, 'base64').length : 0,
                      message: body.Message });
      if (verbose) console.error('[kms-stub]', target, JSON.stringify(body).slice(0, 160));
      if (req.method !== 'POST' || req.url !== '/') {
        return fail(res, 'SerializationException', `unexpected ${req.method} ${req.url}`);
      }
      switch (target) {
        case 'TrentService.GetPublicKey':
          return getPublicKey(res, body);
        case 'TrentService.Sign':
          return sign(res, body);
        default:
          return fail(res, 'UnsupportedOperationException', `unsupported target '${target}'`);
      }
    });
  });

  return {
    server,
    get requests() {
      return requests;
    },
    reset() {
      requests = [];
    },
    /* Which specs this stub can actually serve, so tests can skip the rest. */
    supported() {
      const out = {};
      for (const [spec, { algorithm, cliName }] of Object.entries(SPECS)) {
        out[spec] = cliName ? opensslSupports(cliName) : true;
        if (out[spec]) {
          try {
            generateKeyPairSync(algorithm, SPECS[spec].options);
          } catch {
            out[spec] = false;
          }
        }
      }
      return out;
    },
    async listen() {
      await new Promise((r) => server.listen(0, '127.0.0.1', r));
      const { port } = server.address();
      return { port, endpoint: `http://127.0.0.1:${port}` };
    },
    async close() {
      await new Promise((r) => server.close(r));
      rmSync(dir, { recursive: true, force: true });
    },
  };
}

if (process.argv[1] && import.meta.url.endsWith(process.argv[1].replace(/^.*\//, ''))) {
  const stub = createKmsStub({ verbose: true });
  const { port, endpoint } = await stub.listen();
  console.log(`STUB_PORT=${port}`);
  console.log(`AWS_ENDPOINT_URL_KMS=${endpoint}`);
}

File: src/kms_stub.c
/*
 * A local stand-in for AWS KMS: no network, no credentials, no SDK.
 *
 * This exists so the provider can be developed and tested offline, and so the
 * store/keymgmt/signature machinery can be exercised without an AWS account. It
 * is NOT a production backend and is only compiled when AWSKMS_BACKEND=stub.
 *
 * The key spec is taken from the key id: any KMS KeySpec name appearing in it
 * selects that spec, so realistic URIs work --
 *   awskms:key-id=alias/my-RSA_2048-signer
 *   awskms:key-id=arn:aws:kms:eu-central-1:111122223333:key/ECC_NIST_P256
 * A key pair is generated on first use and cached under the key id for the life
 * of the process, so a load, a sign and a verify all agree with each other.
 */
#include <openssl/core_names.h>
#include <openssl/crypto.h>
#include <openssl/evp.h>
#include <openssl/rsa.h>
#include <openssl/x509.h>
#include <string.h>

#include "err.h"
#include "keyspec.h"
#include "kms.h"

const char *awskms_kms_backend(void) { return "stub"; }

void awskms_public_key_cleanup(AWSKMS_PUBLIC_KEY *p) {
  if (p == NULL) return;
  OPENSSL_free(p->key_spec);
  OPENSSL_free(p->spki);
  memset(p, 0, sizeof(*p));
}

/* ------------------------------------------------------------ the fake token */

struct entry {
  char *key_id;
  const AWSKMS_KEYSPEC *spec;
  EVP_PKEY *pkey;
  struct entry *next;
};

static struct entry *entries;
static CRYPTO_ONCE once = CRYPTO_ONCE_STATIC_INIT;
static CRYPTO_RWLOCK *lock;

static void init_lock(void) { lock = CRYPTO_THREAD_lock_new(); }

/* Case-insensitive search for a KeySpec name anywhere in the key id. */
static const AWSKMS_KEYSPEC *spec_from_key_id(const char *key_id) {
  for (const AWSKMS_KEYSPEC *s = awskms_keyspec_all(); s->kms_key_spec != NULL;
       s++) {
    size_t n = strlen(s->kms_key_spec);
    for (const char *p = key_id; *p != '\0'; p++)
      if (OPENSSL_strncasecmp(p, s->kms_key_spec, n) == 0) return s;
  }
  return NULL;
}

static EVP_PKEY *generate(const AWSKMS_KEYSPEC *spec) {
  switch (spec->family) {
    case AWSKMS_FAMILY_RSA: {
      size_t bits = 2048;
      if (strcmp(spec->kms_key_spec, "RSA_3072") == 0) bits = 3072;
      if (strcmp(spec->kms_key_spec, "RSA_4096") == 0) bits = 4096;
      return EVP_PKEY_Q_keygen(NULL, NULL, "RSA", bits);
    }
    case AWSKMS_FAMILY_EC:
      return EVP_PKEY_Q_keygen(NULL, NULL, "EC", spec->group_name);
    case AWSKMS_FAMILY_ED25519:
      return EVP_PKEY_Q_keygen(NULL, NULL, "ED25519");
    case AWSKMS_FAMILY_ML_DSA:
      return EVP_PKEY_Q_keygen(NULL, NULL, spec->keytype);
  }
  return NULL;
}

/* Returns a borrowed entry, generating it on first use. */
static struct entry *lookup(AWSKMS_PROV_CTX *provctx, const char *key_id) {
  const AWSKMS_KEYSPEC *spec;
  struct entry *e;

  CRYPTO_THREAD_run_once(&once, init_lock);
  if (!CRYPTO_THREAD_write_lock(lock)) return NULL;

  for (e = entries; e != NULL; e = e->next)
    if (strcmp(e->key_id, key_id) == 0) goto out;

  if ((spec = spec_from_key_id(key_id)) == NULL) {
    AWSKMS_raise(provctx->handle, AWSKMS_R_KEY_NOT_FOUND,
                 "stub backend: key id \"%s\" names no KMS key spec", key_id);
    goto out;
  }

  if ((e = OPENSSL_zalloc(sizeof(*e))) == NULL) goto out;
  if ((e->key_id = OPENSSL_strdup(key_id)) == NULL ||
      (e->pkey = generate(spec)) == NULL) {
    AWSKMS_raise(provctx->handle, AWSKMS_R_INTERNAL_ERROR,
                 "stub backend: could not generate a %s key",
                 spec->kms_key_spec);
    OPENSSL_free(e->key_id);
    EVP_PKEY_free(e->pkey);
    OPENSSL_free(e);
    e = NULL;
    goto out;
  }
  e->spec = spec;
  e->next = entries;
  entries = e;

out:
  CRYPTO_THREAD_unlock(lock);
  return e;
}

/* ------------------------------------------------------------ fault injection */

/*
 * A key id containing one of these markers makes GetPublicKey return something
 * malformed, which is what exercises the provider's own validation.
 *
 * Those paths are otherwise unreachable offline. They guard against a key spec
 * this provider does not implement (SM2, HMAC) and against a SubjectPublicKeyInfo
 * that does not match the KeySpec reported alongside it -- both producible only by
 * the real service. A marker in the key id is the least invasive way to reach
 * them, and is test-only by construction, since this file is compiled for the stub
 * backend alone.
 */
static int has_fault(const char *key_id, const char *name) {
  return strstr(key_id, name) != NULL;
}

int awskms_kms_get_public_key(AWSKMS_PROV_CTX *provctx, const AWSKMS_URI *uri,
                              AWSKMS_PUBLIC_KEY *out) {
  struct entry *e = lookup(provctx, uri->key_id);
  unsigned char *der = NULL;
  int len;

  if (e == NULL) return 0;

  if ((len = i2d_PUBKEY(e->pkey, &der)) <= 0) {
    AWSKMS_raise(provctx->handle, AWSKMS_R_GET_PUBLIC_KEY_FAILED,
                 "stub backend: could not encode the public key");
    return 0;
  }
  /* A KeySpec this provider does not implement -- what a real SM2 or HMAC key
   * would report. Must surface as UNSUPPORTED_KEY_SPEC at load. */
  if (has_fault(uri->key_id, "fault-badspec")) {
    out->key_spec = OPENSSL_strdup("SM2");
  } else if (has_fault(uri->key_id, "fault-wrongtype")) {
    /* Claim RSA while serving whatever the key id actually named. Catches the
     * KeySpec/SPKI disagreement in EVP_PKEY_is_a. */
    out->key_spec = OPENSSL_strdup("RSA_2048");
  } else if (has_fault(uri->key_id, "fault-wronggroup")) {
    /* Right key type, wrong curve -- the one a lazy is_a() check would miss. */
    out->key_spec = OPENSSL_strdup(
        strcmp(e->spec->kms_key_spec, "ECC_NIST_P256") == 0 ? "ECC_NIST_P384"
                                                            : "ECC_NIST_P256");
  } else {
    out->key_spec = OPENSSL_strdup(e->spec->kms_key_spec);
  }
  if (out->key_spec == NULL) {
    OPENSSL_free(der);
    return 0;
  }

  /* SPKI that does not match the KeySpec alongside it. Each must surface as
   * MALFORMED_PUBLIC_KEY rather than as a key that half-works. */
  if (has_fault(uri->key_id, "fault-emptyspki")) {
    OPENSSL_free(der);
    out->spki = NULL;
    out->spki_len = 0;
    return 1;
  }
  if (has_fault(uri->key_id, "fault-truncspki")) {
    out->spki = der;
    out->spki_len = (size_t)len / 2;
    return 1;
  }
  if (has_fault(uri->key_id, "fault-badspki")) {
    /* Corrupt the body but leave the leading SEQUENCE tag, so this fails in the
     * parser rather than being rejected as obviously-not-DER. */
    for (int i = len / 2; i < len; i++) der[i] ^= 0xff;
    out->spki = der;
    out->spki_len = (size_t)len;
    return 1;
  }

  /* i2d_PUBKEY allocates with OPENSSL_malloc, so ownership transfers cleanly. */
  out->spki = der;
  out->spki_len = (size_t)len;
  return 1;
}

/* Mirrors KMS's own semantics for a SigningAlgorithm: which digest it implies and
 * whether it is PSS. */
static int parse_algorithm(const char *algorithm, const char **md_name,
                           int *pss) {
  const char *suffix;

  *pss = 0;
  if ((suffix = strstr(algorithm, "_SHA_")) != NULL) {
    suffix += 5;
    if (strcmp(suffix, "256") == 0)
      *md_name = "SHA2-256";
    else if (strcmp(suffix, "384") == 0)
      *md_name = "SHA2-384";
    else if (strcmp(suffix, "512") == 0)
      *md_name = "SHA2-512";
    else
      return 0;
    *pss = strncmp(algorithm, "RSASSA_PSS_", 11) == 0;
    return 1;
  }
  return 0;
}

int awskms_kms_sign(AWSKMS_PROV_CTX *provctx, const AWSKMS_URI *uri,
                    const char *signing_algorithm, const char *message_type,
                    const unsigned char *msg, size_t msg_len,
                    unsigned char *sig, size_t sig_size, size_t *sig_len) {
  struct entry *e = lookup(provctx, uri->key_id);
  EVP_PKEY_CTX *pctx = NULL;
  EVP_MD *md = NULL;
  const char *md_name = NULL;
  int pss = 0;
  size_t out_len = sig_size;
  int ok = 0;

  if (e == NULL) return 0;

  /* KMS's documented Message limit. The provider is expected to have rejected
   * oversize input already, so reaching this means a bug upstream of here. */
  if (msg_len < 1 || msg_len > 4096) {
    AWSKMS_raise(provctx->handle, AWSKMS_R_SIGN_FAILED,
                 "stub backend: Message must be 1-4096 bytes, got %zu",
                 msg_len);
    return 0;
  }

  /* MessageType=RAW means "sign the message itself", which for Ed25519 is a
   * one-shot EVP_DigestSign with no digest -- exactly what KMS does. */
  if (strcmp(message_type, "RAW") == 0) {
    EVP_MD_CTX *mdctx = EVP_MD_CTX_new();
    size_t raw_len = sig_size;

    if (mdctx != NULL &&
        EVP_DigestSignInit_ex(mdctx, NULL, NULL, NULL, NULL, e->pkey, NULL) ==
            1 &&
        EVP_DigestSign(mdctx, sig, &raw_len, msg, msg_len) == 1) {
      *sig_len = raw_len;
      ok = 1;
    }
    EVP_MD_CTX_free(mdctx);
    if (!ok)
      AWSKMS_raise(provctx->handle, AWSKMS_R_SIGN_FAILED,
                   "stub backend: RAW signing with %s failed",
                   signing_algorithm);
    return ok;
  }

  /*
   * MessageType=EXTERNAL_MU: the Message already IS the 64-byte FIPS 204 mu, and
   * KMS skips the public-key/message hashing it would normally do. The matching
   * OpenSSL operation is EVP_PKEY_sign_message_init() with the "mu" parameter
   * set -- verified to be the only route that works: plain EVP_PKEY_sign_init and
   * EVP_PKEY_sign_init_ex2 both fail for ML-DSA on 3.5 and 3.6, and the resulting
   * signature verifies as ordinary pure ML-DSA.
   *
   * EVP_PKEY_sign_message_init is an OpenSSL 3.5 linker symbol, so it is guarded.
   * That costs nothing: ML-DSA does not exist before 3.5 either, so on an older
   * host there is no ML-DSA key to sign. (Only this test-only backend references
   * it; the aws backend, which is what ships, does not.)
   */
  if (strcmp(message_type, "EXTERNAL_MU") == 0) {
#if OPENSSL_VERSION_NUMBER >= 0x30500000L
    EVP_SIGNATURE *alg = NULL;
    int one = 1;
    OSSL_PARAM params[2];

    if (msg_len != 64) {
      AWSKMS_raise(
          provctx->handle, AWSKMS_R_SIGN_FAILED,
          "stub backend: EXTERNAL_MU requires a 64-byte Message, got %zu",
          msg_len);
      return 0;
    }

    params[0] = OSSL_PARAM_construct_int(OSSL_SIGNATURE_PARAM_MU, &one);
    params[1] = OSSL_PARAM_construct_end();

    alg = EVP_SIGNATURE_fetch(NULL, e->spec->keytype, NULL);
    if (alg != NULL &&
        (pctx = EVP_PKEY_CTX_new_from_pkey(NULL, e->pkey, NULL)) != NULL &&
        EVP_PKEY_sign_message_init(pctx, alg, params) > 0 &&
        EVP_PKEY_sign(pctx, sig, &out_len, msg, msg_len) > 0) {
      *sig_len = out_len;
      ok = 1;
    }
    EVP_SIGNATURE_free(alg);
    EVP_PKEY_CTX_free(pctx);
    if (!ok)
      AWSKMS_raise(provctx->handle, AWSKMS_R_SIGN_FAILED,
                   "stub backend: EXTERNAL_MU signing with %s failed",
                   signing_algorithm);
    return ok;
#else
    AWSKMS_raise(
        provctx->handle, AWSKMS_R_SIGN_FAILED,
        "stub backend: EXTERNAL_MU needs OpenSSL 3.5 headers, which is "
        "also the floor for ML-DSA");
    return 0;
#endif
  }

  if (strcmp(message_type, "DIGEST") != 0) {
    AWSKMS_raise(provctx->handle, AWSKMS_R_SIGN_FAILED,
                 "stub backend: MessageType %s is not implemented",
                 message_type);
    return 0;
  }
  if (!parse_algorithm(signing_algorithm, &md_name, &pss)) {
    AWSKMS_raise(provctx->handle, AWSKMS_R_SIGN_FAILED,
                 "stub backend: unknown SigningAlgorithm %s",
                 signing_algorithm);
    return 0;
  }

  /* EVP_PKEY_sign on a pre-computed digest is exactly what MessageType=DIGEST
   * means, so the stub and the real service agree by construction. */
  if ((md = EVP_MD_fetch(NULL, md_name, NULL)) == NULL) goto out;
  if ((pctx = EVP_PKEY_CTX_new_from_pkey(NULL, e->pkey, NULL)) == NULL)
    goto out;
  if (EVP_PKEY_sign_init(pctx) <= 0) goto out;
  if (e->spec->family == AWSKMS_FAMILY_RSA) {
    int padding = pss ? RSA_PKCS1_PSS_PADDING : RSA_PKCS1_PADDING;
    if (EVP_PKEY_CTX_set_rsa_padding(pctx, padding) <= 0) goto out;
    /* KMS always uses a salt length equal to the digest length. */
    if (pss &&
        EVP_PKEY_CTX_set_rsa_pss_saltlen(pctx, RSA_PSS_SALTLEN_DIGEST) <= 0)
      goto out;
  }
  if (EVP_PKEY_CTX_set_signature_md(pctx, md) <= 0) goto out;
  if (EVP_PKEY_sign(pctx, sig, &out_len, msg, msg_len) <= 0) goto out;

  *sig_len = out_len;
  ok = 1;

out:
  if (!ok)
    AWSKMS_raise(provctx->handle, AWSKMS_R_SIGN_FAILED,
                 "stub backend: signing with %s failed", signing_algorithm);
  EVP_PKEY_CTX_free(pctx);
  EVP_MD_free(md);
  return ok;
}
